File: 0079-word-search/0079-word-search.cpp
class Solution {
public:
    bool depthFirstSearch(vector<vector<char>> & board, string word, int i, int j, int count){
        if(count == word.size()){
            return true;
        }
        if(i < 0 || i >= board.size() || j<0 || j >= board[0].size() || word[count] != board[i][j]){
            return false;
        }
        char temp = board[i][j];
        board[i][j] = '*';
        bool isExist = depthFirstSearch(board, word, i+1, j, count+1) || depthFirstSearch(board, word, i-1, j, count+1) || depthFirstSearch(board, word, i, j+1, count+1) || depthFirstSearch(board, word, i, j-1, count+1);
        board[i][j] = temp;
        return isExist;
    }
    
    bool exist(vector<vector<char>>& board, string word) {
        for(int i = 0; i < board.size(); i++){
            for(int j = 0; j < board[0].size(); j++){
                if(board[i][j] == word[0] && depthFirstSearch(board, word, i, j, 0)){
                    return true;
                }
            }
        }
        return false;
    }
};

File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        if not board: return False
        for i in range(len(board)):
            for j in range(len(board[0])):
                if self.depthFirstSearch( board, i, j, word):
                    return True
        return False
    def depthFirstSearch(self, board, i, j, word):
        if len(word)==0: return True
        if i<0 or i>= len(board) or j<0 or j>= len(board[0]) or word[0] != board[i][j]: return False
        
        temp = board[i][j]
        board[i][j] = '*'
        isExist = self.depthFirstSearch(board, i+1, j, word[1:]) or self.depthFirstSearch(board, i-1, j, word[1:]) or self.depthFirstSearch(board, i, j+1, word[1:]) or self.depthFirstSearch(board, i, j-1, word[1:])
        board[i][j] = temp
        return isExist
    